sql/select.go

```diff
@@ -85,3 +99,523 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
 	parser.WalkStmt(&v, stmt)
 	return v.err
 }
+
+// selectIndex analyzes the scanNode to determine if there is an index
+// available that can fulfill the query with a more restrictive scan.
+//
+// The current occurs in two passes. The first pass performs a limited form of
+// value range propagation for the qvalues (i.e. the columns). The second pass
+// takes the value range information and determines which indexes can fulfill
+// the query (we currently only support covering indexes) and selects the
+// "best" index from that set. The cost model based on keys per row, key size
+// and number of index elements used.
+func (p *planner) selectIndex(s *scanNode) (planNode, error) {
+	if s.desc == nil || s.filter == nil {
+		// No table or where-clause.
+		return s, nil
+	}
+
+	rangeInfo := make(qvalueRangeMap)
+	rangeInfo.analyzeExpr(s.filter)
+
+	indexes := make([]*indexInfo, 0, len(s.desc.Indexes)+1)
+	indexes = append(indexes, &indexInfo{
+		desc:  s.desc,
+		index: &s.desc.PrimaryIndex,
+	})
+	for i := range s.desc.Indexes {
+		indexes = append(indexes, &indexInfo{
+			desc:  s.desc,
+			index: &s.desc.Indexes[i],
+		})
+	}
+
+	for _, index := range indexes {
+		index.analyzeRanges(s, rangeInfo)
+	}
+
+	sort.Sort(indexInfoByCost(indexes))
+
+	if log.V(2) {
+		for i, index := range indexes {
+			log.Infof("%d: selectIndex(%s): %v %s %s",
+				i, index.index.Name, index.cost, index.makeStartKey(), index.makeEndKey())
+		}
+	}
+
+	if s.isSecondaryIndex {
+		// An explicit secondary index was requested. Keep using it, but use the
+		// start and end keys computed during index selection.
+		for _, index := range indexes {
+			if s.index == index.index {
+				s.startKey = index.makeStartKey()
+				s.endKey = index.makeEndKey()
+				return s, nil
+			}
+		}
+	}
+
+	// After sorting, indexes[0] contains the best index. Copy its info into the
```

Comment:
`indexes[0].index` is a `*IndexDescriptor` which points into the `TableDescriptor`. So yes, I meant to copy the pointer.